### PR TITLE
fix: enable string replace dependency

### PR DESCRIPTION
**Pull request by underfin**
## Related issue (if exists)

closed https://github.com/web-infra-dev/rspack/issues/3056, https://github.com/web-infra-dev/rspack/issues/2190 

## Summary

The pr enable string replace dependency at codegen instead of ast path transform, it also migrate webpack interop test and some bug fix. Here also some features is not aligin to webpack, pelase see https://github.com/web-infra-dev/rspack/issues/3553.

### CODE REVIEW

**Comment by underfin** [comment]
!bench

**Comment by underfin** [comment]
!bench

File: crates/rspack/tests/samples/remove-parent-modules/simple/expected/a_js.js
Comment (underfin) [inline crates/rspack/tests/samples/remove-parent-modules/simple/expected/a_js.js:4]: The snapshot changed becuase the custom runtime is not same as swc commonjs

File: crates/cargo-rst/src/rst.rs
Comment (hardfist) [inline crates/cargo-rst/src/rst.rs:353]: same as https://github.com/mitsuhiko/insta/pull/172/files

**Comment by Boshen** [comment]
@web-infra-dev/rspack-core We need at least 5 review approvals before merging this.

**Comment by hardfist** [comment]
check this https://github.com/web-infra-dev/rspack/issues/2927

File: examples/arco-pro/rspack.config.js
Comment (hardfist) [inline examples/arco-pro/rspack.config.js:15]: why you change this

File: crates/rspack_core/src/runtime_globals.rs
Comment (hardfist) [inline crates/rspack_core/src/runtime_globals.rs:239]: why remove this

File: crates/rspack_loader_sass/tests/fixtures/rspack_importer/expected/main.css
Comment (hardfist) [inline crates/rspack_loader_sass/tests/fixtures/rspack_importer/expected/main.css:5]: this seems unexpected

File: crates/rspack_plugin_css/src/dependency/import.rs
Comment (hardfist) [inline crates/rspack_plugin_css/src/dependency/import.rs:72]: what does -8 mean here?

File: crates/rspack_plugin_css/tests/fixtures/custom/http_import/expected/main.css
Comment (hardfist) [inline crates/rspack_plugin_css/tests/fixtures/custom/http_import/expected/main.css:3]: seem unexpected

File: crates/rspack_plugin_css/tests/fixtures/webpack/at-charset/expected/main.css
Comment (hardfist) [inline crates/rspack_plugin_css/tests/fixtures/webpack/at-charset/expected/main.css:26]: +1 here

File: crates/rspack_plugin_css/tests/fixtures/webpack/shared-import/expected/main.css
Comment (hardfist) [inline crates/rspack_plugin_css/tests/fixtures/webpack/shared-import/expected/main.css:10]: +1

File: crates/rspack_plugin_css/tests/fixtures/webpack/at-import/expected/main.css
Comment (hardfist) [inline crates/rspack_plugin_css/tests/fixtures/webpack/at-import/expected/main.css:20]: +1

File: crates/rspack_plugin_css/tests/fixtures/webpack/at-import-in-the-middle/expected/main.css
Comment (hardfist) [inline crates/rspack_plugin_css/tests/fixtures/webpack/at-import-in-the-middle/expected/main.css:21]: +1

File: crates/rspack_core/src/runtime_globals.rs
Comment (underfin) [reply crates/rspack_core/src/runtime_globals.rs:239]: It is swc herlps, we don't need this at now

File: crates/rspack_plugin_css/src/dependency/import.rs
Comment (underfin) [reply crates/rspack_plugin_css/src/dependency/import.rs:72]: because the swc bugs https://github.com/swc-project/swc/issues/7584, so here use `@import ` length to get the origin decl start

File: crates/rspack_loader_sass/tests/fixtures/rspack_importer/expected/main.css
Comment (underfin) [reply crates/rspack_loader_sass/tests/fixtures/rspack_importer/expected/main.css:5]: This always cause by https://github.com/swc-project/swc/issues/7584, the end of decl is not correct.

File: crates/rspack_plugin_css/tests/fixtures/webpack/at-charset/expected/main.css
Comment (underfin) [reply crates/rspack_plugin_css/tests/fixtures/webpack/at-charset/expected/main.css:26]: This always cause by https://github.com/swc-project/swc/issues/7584, the end of decl is not correct.

**Comment by Boshen** [comment]
@suxin2017 I want a review approval from you as well, if you have the time.

File: crates/rspack_plugin_javascript/src/dependency/context/common_js_require_context_dependency.rs
Comment (ahabhgk) [inline crates/rspack_plugin_javascript/src/dependency/context/common_js_require_context_dependency.rs:69]: can `Box` be removed?

File: crates/rspack_plugin_javascript/src/dependency/context/import_context_dependency.rs
Comment (ahabhgk) [inline crates/rspack_plugin_javascript/src/dependency/context/import_context_dependency.rs:69]: Same here

File: crates/rspack/tests/tree-shaking/array-side-effects/expected/main.js
Comment (underfin) [inline crates/rspack/tests/tree-shaking/array-side-effects/expected/main.js:3]: The tree shaking test has test script for diff, it can be reviewed by @IWANABETHATGUY, others could skip it.

**Comment by ahabhgk** [comment]
!eco-ci

**Review by hardfist — DISMISSED** [review]

**Review by hyf0 — APPROVED** [review]

**Comment by underfin** [comment]
!bench

**Review by h-a-n-a — APPROVED** [review]

**Review by IWANABETHATGUY — APPROVED** [review]

**Review by ahabhgk — APPROVED** [review]

**Review by hardfist — APPROVED** [review]

**Pull request merged**
